cylc/flow/subprocpool.py

```diff
@@ -384,17 +436,112 @@ def _run_command_init(cls, ctx, callback=None, callback_args=None):
+                error - usually associated with ssh being unable to
+                contact a remote host.
+            callback_255_args: Arguments for the 255 callback function.
+
+        """
+        def _run_callback(callback, args_=None):
+            if callable(callback):
+                if not args_:
+                    args_ = []
+                callback(ctx, *args_)
+            else:
+                return False
+
         ctx.timestamp = get_current_time_string()
-        if callable(callback):
-            if not callback_args:
-                callback_args = []
-            callback(ctx, *callback_args)
+
+        if cls.ssh_255_fail(ctx) or cls.rsync_255_fail(ctx) is True:
+            # Job log retrieval passes a special object as a command key
+            # Extra logic to provide sensible strings for logging.
+            if isinstance(ctx.cmd_key, TaskJobLogsRetrieveContext):
+                cmd_key = ctx.cmd_key.key
+            else:
+                cmd_key = ctx.cmd_key
+            LOG.warning(
+                f'"{cmd_key}" failed because "{ctx.host}" is not '
+                f'available right now. "{ctx.host}" has been added to the '
+                f'list of unreachable hosts and {cmd_key} will retry '
+                'if another host is available.'
+            )
+
+            # If callback_255 takes the same args as callback, we don't
+            # want to spec those args:
+            if callable(callback_255) and callback_255_args is None:
+                callback_255_args = callback_args
+
+            # Run Callback
+            if bad_hosts is not None:
+                bad_hosts.add(ctx.host)
+            res = _run_callback(callback_255, callback_255_args)
+            if res is False:
+                _run_callback(callback, callback_args)
+        else:
+            # For every other return code run default callback.
+            _run_callback(callback, callback_args)
+
+    @staticmethod
+    def ssh_255_fail(ctx) -> bool:
+        """Test context for ssh command failing with a 255 error."""
+        ssh_255_fail = False
+        if (
+            ctx.cmd[0] == 'ssh'
+            and ctx.ret_code == 255
+        ):
+            ssh_255_fail = True
+        return ssh_255_fail
+
+    @staticmethod
+    def rsync_255_fail(ctx) -> bool:
+        """Tests context for rsync failing to communicate with a host.
+
+        If there has been a failure caused by rsync being unable to connect
+        try a test of ssh connectivity. Necessary becuase loss of connectivity
+        may cause different rsync failures depending on version, and some of
+        the failures may be caused by other problems.
+        """
+        rsync_255_fail = False
+        if (
+            ctx.cmd[0] == 'rsync'
+            and ctx.ret_code not in [0, 255]
+            and is_remote_host(ctx.host)
+        ):
+            ssh_test = run(
+                ['ssh', ctx.host, 'echo', 'True'], capture_output=True
```

Comment:
Slightly more minimal SSH cmd:

```suggestion
                ['ssh', ctx.host, 'true'], capture_output=True
```

Reply:
We should really be using the platform configured SSH command here, but I'm not sure how easy it is to get that with the information we have?

Reply:
> We should really be using the platform configured SSH command here, but I'm not sure how easy it is to get that with the information we have?

Not easy. Can I punt it into another issue?

Reply:
Ok, can you open another issue for that, we will have to tackle it soon is it could cause this SSH to fail incorrectly.

Reply:
I have a fix, I think. Just updating the tests, then Will push·

Reply:
I didn't have a complete fix, it only worked to installation but not log retrieval. Will continue poking it, but reverted the changes so that this branch can be merged.

Pushed into issue #4184